### DISCUSSION TITLE
Random text (obtained from front matter) when browser refreshes

### DISCUSSION THREAD

**Discussion by ariyanna3** [11ty/eleventy#2314] [Q&A] [answered]
Hi everyone,

I'm no developer, so can't figure out what could be the best way to achieve it. The issue is as described below:

1. I have a text quote in the front matter yaml section of every blog post. Because it's easier to read through the blog post and find some quote from the post after reading through it.
2. I have a section on my home page where I can display a random quote from all those quotes sitting inside blog post md files as described above.
3.   This random quote on the homepage needs to change/replaced with another quote every time a user refreshes the web browser

What is the best way to do it? What I have now doesn't change when browser refreshes but only after rebuilding the entire site.
I learned that Javascript can dynamically show random text when the browser refreshes, but I don't know how I can retrieve my quotes in blog post front matter into a Javascript array ?!? Please help.

Thank you.

**Comment by pdehaan** [comment, marked as answer]
Here's one possible solution...
```sh
tree -a -I node_modules/ -I www/
.
├── .eleventy.js
├── package-lock.json
├── package.json
└── src
    ├── _quotes.liquid
    ├── blog
    │   ├── blog.11tydata.json
    │   ├── five.md
    │   ├── four.md
    │   ├── one.md
    │   ├── three.md
    │   └── two.md
    └── index.liquid

2 directories, 11 files
```

Each of my blog posts in ./src/blog/*.md look roughly like this:

```liquid
---
title: Page 1
quote: Best of times, "less" best of times.
---

<h1>{{ title }}</h1>
<blockquote>
  <p>{{ quote }}</p>
</blockquote>
```

Next, my .eleventy.js creates a custom collection based on the blog posts and tries it's best to extract the `quote` front matter data and return an array of quotes. This would be a bit more complex depending if you needed attributions and citations or not.

```js
// .eleventy.js
module.exports = function (eleventyConfig) {
  eleventyConfig.addCollection("quotes", collectionAPI => {
    const quoteArr = collectionAPI.getFilteredByGlob("./src/blog/*.md");
    return quoteArr.map(q => q.data.quote).filter(Boolean);
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

Basically we're creating a custom collection based off the files in the ./src/blog/*.md directory. Next we use `.map()` to extract _just_ the `quote` attribute from the front matter, and finally `.filter()` to remove any empty array elements (in case some blog posts don't have a `quote`, it shouldn't return empty strings or undefined/null or whatever.

Now, armed with a custom collection which is an array of strings, we can create a template which outputs the quotes as a big JSON object.

```liquid
---
# src/_quotes.js
permalink: /js/quotes.js
layout: ""
---

const quotes = {{ collections.quotes | json }};
```

First we set the `permalink` to an &lt;outputDir&gt;/js/quotes.js file, turn off any inherited layout file, then set a `quotes` variable with the JSON-ified version of the quote array. If you're using Nunjucks instead of Liquid templates, the filter name will be [`dump`](https://mozilla.github.io/nunjucks/templating.html#dump), I believe.

Finally, we have our homepage which will set some HTML placeholder for a random quote, then import the /js/quotes.js file, add some little function to grab a random array element, then set the dynamic quote on the homepage using some client side JavaScript (since if we did it at build time, it'd be the same static quote until you redeployed the site.

```liquid
---
# src/index.liquid
title: Home page
---

<h1>{{ title }}</h1>
<p id="quote"></p>

<script src="/js/quotes.js"></script>
<script>
  function random(arr = []) {
    const idx = Math.floor(Math.random() * arr.length);
    return arr[idx];
  }
  document.getElementById("quote").textContent = random(quotes);
</script>
```

**Reply by ariyanna3** [reply]
@pdehaan, thank you, this looks so promising. I'll try it out now. I should be, obviously, much slower than a developer. But, I'll get back once I'm done. Thanks again.

**Reply by ariyanna3** [reply]
I tested it and it works like a charm :) Thank you very much!

**Reply by pdehaan** [reply]
I will add one word of caution with my solution I thought of last night...

If your site/blog has say, 5000 posts, and each post has a quote, you're potentially loading a _very_ big .js or .json file on the homepage. Make sure you do some checks on your final site to make sure the /js/quotes.js file isn't offensively large and causing site performance issues.
One solution might be to limit the maximum number of available quotes.
Using something like https://lodash.com/docs/4.17.15#sampleSize in your src/_quotes.js might be a quick way of randomizing all the quotes and limiting the homepage to, say, 25-50 possible quotes. Each time you redeploy the site, you'd get a different pool of available quotes, but it might keep that /js/quotes.js file to a much more reasonable size.

Another option might be to sort the `collections.quotes` collection by descending date and then limit the results, so you're only selecting quotes from within the 50 most recent blog posts.

**Comment by pdehaan** [comment]
For my own future reference, [alpine.js](https://codewithhugo.com/alpinejs-x-data-fetching/) is super awesome, and this seems to work for dynamically fetching a JSON file, plucking a random array element, and setting text using data binding:

```js
<div x-data="getQuotes" x-init="fetchQuotes">
  <p x-text="quote"></p>
</div>

<script>
  function getQuotes() {
    return {
      quote: "",
      fetchQuotes() {
        fetch("./js/quotes.json")
          .then(res => res.json())
          .then(data => {
            const idx = Math.floor(Math.random() * data.length);
            this.quote = data[idx];
        });
      }
    }
  }
</script>
<script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
```
